### PR TITLE
BaseConnector and Connector Plugin Implementation

### PR DESCRIPTION
**Pull request by njokuScript**
## Overview
This PR introduces the BaseConnector abstract contract and the Connector plugin system, providing a flexible and extensible architecture for integrating various functionalities into our smart contract ecosystem.

## Key Changes
1. Implemented `BaseConnector` abstract contract
2. Created `IConnector` interface
3. Developed Connector plugin system
4. Updated state mutability in `IConnector` interface

### BaseConnector
The `BaseConnector` abstract contract serves as a foundation for all connector implementations. It includes:
- Basic connector information (name and version)
- Implementation of `getName()` and `getVersion()` functions
- Abstract `execute()` function to be implemented by specific connectors

### IConnector Interface
The `IConnector` interface defines the standard methods that all connectors must implement:
- `getName()`: Returns the connector's name
- `getVersion()`: Returns the connector's version
- `execute()`: Executes connector-specific logic

### Connector Plugin System
The Connector plugin system allows for modular and extensible functionality:
- Enables easy integration of new features through connectors
- Provides a standardized way to interact with different connectors
- Allows for future upgrades and additions without modifying core contracts

### State Mutability Update
- Changed `getName()` and `getVersion()` in `IConnector` from `pure` to `view` to align with `BaseConnector` implementation

## Testing
- Added unit tests for `BaseConnector` functionality
- Implemented integration tests for the Connector plugin system

## Security Considerations
- Ensured that the `execute()` function in connectors cannot access unauthorized contract state
- Implemented access control measures for connector management

## Next Steps
- Develop specific connector implementations (e.g., Liquidity protocols, swap protocols and cross-chain bridges)
- Create documentation for connector development and integration

## Feedback Requested
- Review of the overall architecture and extensibility of the connector system
- Suggestions for additional features or improvements to the `BaseConnector`

### CODE REVIEW

**Review by 0xdavinchee — COMMENTED** [review]
looks good overall, some minor questions/suggestions

File: src/BaseConnector.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    constructor(string memory name_, uint256 version_) {
+        _name = keccak256(bytes(name_));
+        _version = version_;
+    }
+
+    /**
+     * @notice Gets the name of the connector
+     * @return string The name of the connector
+     */
+    function getName() external view override returns (string memory) {
```
[the remark this answers is not in the document]
Comment (njokuScript) [reply src/BaseConnector.sol:36]: in the constructor, before storing the name, I converted it to bytes32.

File: test/ConnectorPlugin.t.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        registry.addConnector(address(approvedConnector), "ApprovedConnector", 1);
+    }
+
+    function testExecuteApprovedConnector() public {
+        bytes memory data = abi.encodeWithSignature("mockFunction(uint256)", 42);
+        bytes memory result = plugin.execute(address(approvedConnector), data);
+        assertEq(abi.decode(result, (uint256)), 42);
+    }
+
+    function testExecuteUnapprovedConnector() public {
```
[the remark this answers is not in the document]
Comment (njokuScript) [reply test/ConnectorPlugin.t.sol:32]: got it. will update in the next pr

**Review by 0xdavinchee — APPROVED** [review]
LGTM

**Pull request merged**